src/webgpu/util/f32_interval.ts

```diff
@@ -746,26 +746,47 @@ const AtanIntervalOp: PointToIntervalOp = {
 };
 
 /** Calculate an acceptance interval of atan(x) */
-export function atanInterval(n: number): F32Interval {
+export function atanInterval(n: number | F32Interval): F32Interval {
   return runPointOp(toF32Interval(n), AtanIntervalOp);
 }
 
 const Atan2IntervalOp: BinaryToIntervalOp = {
   impl: (y: number, x: number): F32Interval => {
-    const numULP = 4096;
+    if (x === 0) {
+      // atan2 is not defined meaningfully here
+      return F32Interval.any();
+    }
+
+    // y/x == 0, [-π, π]
     if (y === 0) {
-      if (x === 0) {
-        return F32Interval.any();
-      } else {
-        return F32Interval.span(
-          ulpInterval(kValue.f32.negative.pi.whole, numULP),
-          ulpInterval(kValue.f32.positive.pi.whole, numULP)
-        );
-      }
+      // For x > 0 atan2 is supposed to be 0, and for x < 0 atan2 is -π or π depending on the direction y/x approaches
+      // 0.
+      // Because implementations will be be calculating this via numeric approximations that will vary slightly around
+      // 0 and subnormals, it is very difficult to write a rule for accuracy here without just dictating a specific
+      // implementation. Instead if the value is at y/x = 0, then the result should be in [-π, π], plus 4096 ULP from
+      // atan.
+      return F32Interval.span(
```

Comment:
From the IEEE rules:

* If y is ±0 and x is negative or -0, ±π is returned
* If y is ±0 and x is positive or +0, ±0 is returned

Can we follow that here?   If  y === 0,  and X >0, then expect 0  within 4096 ULPs

Graphically, this follows the middle curve in this graph https://en.wikipedia.org/wiki/Atan2#/media/File:Arctangent2.svg to say that when x >0 then as y/x approaches zero, atan2(y,x) smoothly approaches 0.


Or do you already have evidence that drivers are erratic for  y == 0  and X> 0

Reply:
@amaiorano I believe there were Windows drivers that were returning ±π in unexpected cases, but I don't recall the details.  Do you recall?

Those cases may be covered by `If y is ±0 and x is negative or -0, ±π is returned`, so I will implement what you have sugggested and we can see if it is breaking anywhere.

Reply:
Specifically this is becoming an issue when y is subnormal. Probably something related to y = 0 causing y/x = 0 and that is hard coded or something in the driver.

For example:
    (-1.1754942106924411e-38 (0x807fffff subnormal), 12.734692573547363 (0x414bc14d))
        returned: 3.1415927410125732 (0x40490fdb)
        expected: [-4.81482486096809e-35,4.81482486096809e-35]

Reply:
Hmmm.... it appears that π is being returned unexpected when y is a negative subnormal, so maybe they are just doing -0 wrong?

Reply:
Ok, it's very clarifying that it's the "y is subnormal" case.  The function is wildly approaching discontinuity from -pi to +pi as y transitions from small negative to small positive.
(Stare at the colour diagram at https://en.cppreference.com/w/cpp/numeric/math/atan2 )

Allowing for small errors in calculation it's not particularly meaningful to ask about the quadrant of a point near the origin.

Suggest we update the spec to say the accuracy of atan2 at most when y is normal.